File: Calto-GG/Goldman Sachs/Day 3/IPO.py
# First we sort the projects by their capital requirements in ascending order. 
# Then, use a while loop to iterate over the sorted projects and add the 
# most profitable projects (with capital requirements less than or equal to 
# the current available capital) to a heap. Finally, we repeatedly select the 
# most profitable project from the heap (up to k times) and subtract the project's 
# profit from the current capital until the heap is empty or k projects have been 
# selected. The final capital is returned as the result.

# heapq is used to implement the priority queue. The negative sign before 
# the project's profit is used to make the heap work as a max heap.

import heapq
class Solution(object):
    def findMaximizedCapital(self, k, w, profits, capital):
        projects = sorted(zip(profits, capital), key=lambda x: x[1])
        heap = []
        i = 0
        for _ in range(k):
            while i < len(projects) and projects[i][1] <= w:
                heapq.heappush(heap, -projects[i][0])
                i += 1
            if heap:
                w -= heapq.heappop(heap)
        return w

File: PratapGS/Goldman Sachs/Day3/IPO.cpp



class Solution{
public:
int findMaximizedCapital(int k, int w, vector<int>& profits, vector<int>& capital) {
    
    int n=profits.size();

    vector<pair<int,int>> projects(n); // forming a pair to store profit and capitals

    for(int i=0;i<n;i++)
    
    {
         projects[i]={profits[i],capital[i]}; 
    }

    int i=0;
    // Sorting pairs as per Capital
    sort(projects.begin(),projects.end(),[&](pair<int,int> a,pair<int,int> b){ return a.second<b.second;});

    // to take the max profit from the projects
    priority_queue<int> maxProfit;


    while(k--){
        while(i<n && w>=projects[i].second)
        
        {
            
             maxProfit.push(projects[i++].first);

        }
        
        if(!maxProfit.empty()){
            
             ans+=maxProfit.top(),maxProfit.pop();

        }
    }
    
    return ans;
}
};
